Don't prepare the response body for HEAD requests

anton-vlasenko opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/56481

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

anton-vlasenko committed:
Short circuit the WP_REST_Posts_Controller controller.

anton-vlasenko committed:
Short circuit the WP_REST_Comments_Controller controller.

anton-vlasenko committed:
Short circuit the WP_REST_Terms_Controller controller.

anton-vlasenko committed:
Short circuit the WP_REST_Users_Controller controller.

anton-vlasenko committed:
Short circuit the WP_REST_Taxonomies_Controller controller.

anton-vlasenko committed:
Use the upper-case http method name (per https://core.trac.wordpress.org/ticket/56481#comment:14).

anton-vlasenko committed:
Add PHPUnit tests for the WP_Request::is_method() method.

anton-vlasenko committed:
Better test method name.

ironprogrammer commented on tests/phpunit/tests/rest-api/rest-categories-controller.php:
```diff
@@ -1236,4 +1254,91 @@ protected function check_get_taxonomy_term_response( $response ) {
+	 *
+	 * @param string $method HTTP method to use.
+	 */
+	public function test_get_items_returns_only_fetches_ids_for_head_requests( $method ) {
```
```suggestion
	public function test_get_items_returns_only_ids_for_head_requests( $method ) {
```
For this and other same-named tests in the updated test classes under this PR, `returns_only_fetches` has two verbs, but should only use one for clarity.

Scanning other test classes, "return" is used more extensively than "fetch", so IMHO these new methods should use `returns_only_ids`. What do you think?

anton-vlasenko replied:
Well spotted, @ironprogrammer! Having two verbs here doesn't make sense. This is likely the result of extensive copy-pasting between different test classes. :)

I'm leaning towards `test_get_items_fetches_only_ids_for_head_requests`, because what I meant is that the `get_items()` method fetches only the `*id*` column from the database for HEAD requests, as opposed to fetching all columns for GET requests.

Fixed in 9c91e455805d50251480a465faf0c499e4f37692.

ironprogrammer commented on tests/phpunit/tests/rest-api/rest-comments-controller.php:
```diff
@@ -3386,8 +3478,73 @@ public function test_check_read_post_permission_with_invalid_post_type() {
+	 *
+	 * @param string $method HTTP method to use.
+	 */
+	public function test_get_items_only_fetches_ids_for_head_requests( $method ) {
```
```suggestion
	public function test_get_items_returns_only_ids_for_head_requests( $method ) {
```
Suggestion per previous note.

anton-vlasenko replied:
https://github.com/WordPress/wordpress-develop/pull/7970#discussion_r1876900479

ironprogrammer commented on tests/phpunit/tests/rest-api/rest-posts-controller.php:
```diff
@@ -1826,6 +1940,70 @@ public function test_get_items_pagination_headers() {
+	 *
+	 * @param string $method HTTP method to use.
+	 */
+	public function test_get_items_returns_only_fetches_ids_for_head_requests( $method ) {
```
```suggestion
	public function test_get_items_returns_only_ids_for_head_requests( $method ) {
```
Suggestion per above.

anton-vlasenko replied:
https://github.com/WordPress/wordpress-develop/pull/7970#discussion_r1876900479

ironprogrammer commented on tests/phpunit/tests/rest-api/rest-tags-controller.php:
```diff
@@ -1502,4 +1508,91 @@ protected function check_get_taxonomy_term_response( $response, $id ) {
+	 *
+	 * @param string $method HTTP method to use.
+	 */
+	public function test_get_items_returns_only_fetches_ids_for_head_requests( $method ) {
```
```suggestion
	public function test_get_items_returns_only_ids_for_head_requests( $method ) {
```
Suggestion per above.

anton-vlasenko replied:
https://github.com/WordPress/wordpress-develop/pull/7970#discussion_r1876900479

ironprogrammer commented on tests/phpunit/tests/rest-api/rest-users-controller.php:
```diff
@@ -3100,6 +3159,79 @@ public function data_get_default_data() {
+	 *
+	 * @param string $method HTTP method to use.
+	 */
+	public function test_get_items_returns_only_fetches_ids_for_head_requests( $method ) {
```
```suggestion
	public function test_get_items_returns_only_ids_for_head_requests( $method ) {
```
Suggestion per above.

anton-vlasenko replied:
https://github.com/WordPress/wordpress-develop/pull/7970#discussion_r1876900479

ironprogrammer reviewed (commented):
Thanks for putting this all together, @anton-vlasenko 🙌🏻 I've got a name tweak suggestion for some of the newly added methods to consider.

anton-vlasenko committed:
Rename the test methods.
Context: https://github.com/WordPress/wordpress-develop/pull/7970#discussion_r1876532067
Props @ironprogrammer.

Mamaduka commented:
Fantastic work, @anton-vlasenko!

Let me know how I can help to land this in WP 6.8. I think it would improve editors' loading performance.

anton-vlasenko commented:
> Fantastic work, @anton-vlasenko!

Thank you, @Mamaduka.

> Let me know how I can help to land this in WP 6.8. I think it would improve editors' loading performance.

Thank you for offering your help!
A code review and/or testing report could help with landing this in WordPress 6.8.
This ticket is likely on @TimothyBJacobs's radar, so hopefully, he'll be able to approve and commit it when he has the time, provided there’s no additional feedback from the code review.

swissspidy commented on tests/phpunit/tests/rest-api/rest-categories-controller.php:
```diff
@@ -1236,4 +1254,91 @@ protected function check_get_taxonomy_term_response( $response ) {
+	 *
+	 * @return array
+	 */
+	public function data_readable_http_methods() {
```
Note that data providers should in general be `static`

Same goes for the other files

anton-vlasenko replied:
Well spotted! I keep forgetting to declare them as static.
Fixed in 2e17f2bbf7f1a1b9f943c66f7789b63196406f16.

anton-vlasenko committed:
Data providers should always be declared as static. Props @swisspidy.

TimothyBJacobs commented:
Yeah I'm planning on landing this this weekend.

spacedmonkey commented on src/wp-includes/rest-api/endpoints/class-wp-rest-comments-controller.php:
```diff
@@ -262,6 +262,12 @@ public function get_items( $request ) {
+		$is_head_request = $request->is_method( 'HEAD' );
+		if ( $is_head_request ) {
+			// Force the 'fields' argument. For HEAD requests, only post IDs are required to calculate pagination.
+			$prepared_args['fields'] = 'ids';
```
Should we avoid priming comment meta here as well?

anton-vlasenko replied:
> Should we avoid priming comment meta here as well?

Yes, I think we should avoid priming comment meta here.
Fixed in 294584e3a25df251df985b2c28525ac264b9adff

spacedmonkey reviewed (changes requested):
I think we need to also disable any meta and term cache priming as well, as this results in database queries. 

Also if there are changes to be made for taxononmy endpoint, those changes should also be made for post type endpoint. Otherwise this looks like great work.

anton-vlasenko committed:
Disable priming meta for HEAD requests to improve performance.

anton-vlasenko committed:
Fix CS and typos.

anton-vlasenko committed:
Don't prepare the response body for HEAD requests in WP_Test_REST_Post_Types_Controller.

anton-vlasenko commented:
> I think we need to also disable any meta and term cache priming as well, as this results in database queries.
> 
> Also if there are changes to be made for taxononmy endpoint, those changes should also be made for post type endpoint. Otherwise this looks like great work.

Thank you for the review, @spacedmonkey.
Well spotted.
I've fixed these issues in 294584e3a25df251df985b2c28525ac264b9adff and b80fd51601bb95ea2ddc04ea98b7e4d79f68ac8c.

TimothyBJacobs commented:
I think we have a BC issue with the single item route. Right now, if I use the `rest_prepare_` filters, I can set headers on the returned response object. After these changes, this no longer happens.

I think we might want to move the `is_method( 'HEAD' )` cehcks into the `prepare_item_for_response` methods, and prepare the empty response object there. That way, if someone adds a header, it is maintained. If they add a property, that's fine because it'll get omitted by the server.

anton-vlasenko commented:
> I think we have a BC issue with the single item route. Right now, if I use the `rest_prepare_` filters, I can set headers on the returned response object. After these changes, this no longer happens.
> 
> I think we might want to move the `is_method( 'HEAD' )` cehcks into the `prepare_item_for_response` methods, and prepare the empty response object there. That way, if someone adds a header, it is maintained. If they add a property, that's fine because it'll get omitted by the server.

Thank you for the review, @TimothyBJacobs.  
While I agree that this is a BC break, I tried searching for any plugins that add headers using the `rest_prepare_` filters and couldn’t find any:  
- https://www.wpdirectory.net/search/01JFAMD046JGJRWHWQ1QNMZAW9
- https://www.wpdirectory.net/search/01JFAMDVT9ZWJ0TS45ZG5B1D1E

It could theoretically happen, but I’m not aware of any such cases.  

That said, I also agree that GET and HEAD requests should behave similarly, with the exception that HEAD requests don’t include a response body.  

I’m working on a fix.  

---

jonnynews reviewed (commented):
There are 42 files in the endpoints directory. Some of these are sub classes. But there are other endpoints that are not sub classes of the files edited in this PR. Is this applied all the endpoints that need this change?

anton-vlasenko commented:
> There are 42 files in the endpoints directory. Some of these are sub classes. But there are other endpoints that are not sub classes of the files edited in this PR. Is this applied all the endpoints that need this change?

@jonnynews  No, it hasn’t been applied to all the classes that need editing, as this PR focuses only on the most commonly used REST controllers.
Do you think the other controllers can be addressed in a follow-up PR, or would you prefer to fix everything in a single PR?

jonnynews commented:
> > There are 42 files in the endpoints directory. Some of these are sub classes. But there are other endpoints that are not sub classes of the files edited in this PR. Is this applied all the endpoints that need this change?
> 
> @jonnynews No, it hasn’t been applied to all the classes that need editing, as this PR focuses only on the most commonly used REST controllers. Do you think the other controllers can be addressed in a follow-up PR, or would you prefer to fix everything in a single PR?

In core, we normally do these things one commit.

anton-vlasenko committed:
Move the check for HEAD requests to ::prepare_item_for_response() method to preserve BC.

jonnynews commented:
Some controllers might benefit from this work. 

```
WP_REST_Block_Pattern_Categories_Controller
WP_REST_Block_Types_Controller
WP_REST_Font_Collections_Controller
WP_REST_Global_Styles_Revisions_Controller
WP_REST_Pattern_Directory_Controller
WP_REST_Search_Controller
WP_REST_Sidebars_Controller
WP_REST_Templates_Controller
WP_REST_Widget_Types_Controller
WP_REST_Widgets_Controller
```

To clear, as I am asking, I believe it just a simply copy and paste for most of these endpoints, but correct me if I am wrong.

anton-vlasenko commented:
> Some controllers might benefit from this work.
> 
> ```
> WP_REST_Block_Pattern_Categories_Controller
> WP_REST_Block_Types_Controller
> WP_REST_Font_Collections_Controller
> WP_REST_Global_Styles_Revisions_Controller
> WP_REST_Pattern_Directory_Controller
> WP_REST_Search_Controller
> WP_REST_Sidebars_Controller
> WP_REST_Templates_Controller
> WP_REST_Widget_Types_Controller
> WP_REST_Widgets_Controller
> ```
> 
> To clear, as I am asking, I believe it just a simply copy and paste for most of these endpoints, but correct me if I am wrong.

Yes, I believe it is. PHPUnit tests would need to be added as well (which can also be partially copy-pasted). I'll work on this tomorrow. Thank you for the review.

anton-vlasenko committed:
Short circuit the WP_REST_Block_Pattern_Categories_Controller controller.

anton-vlasenko committed:
Short circuit the WP_REST_Block_Types_Controller controller.

anton-vlasenko committed:
Fix CS.

anton-vlasenko committed:
Add unit tests.

anton-vlasenko committed:
Short circuit the WP_REST_Font_Collections_Controller controller.

anton-vlasenko committed:
Add unit tests.

anton-vlasenko committed:
Short circuit the WP_REST_Revisions_Controller controller.

anton-vlasenko committed:
Add unit tests.

anton-vlasenko committed:
Short circuit the WP_REST_Global_Styles_Revisions_Controller controller.

anton-vlasenko committed:
Short circuit the WP_REST_Global_Styles_Revisions_Controller controller.

anton-vlasenko committed:
Add unit tests.

anton-vlasenko committed:
Fix the failing unit test.

spacedmonkey reviewed (approved):
Great work. This is good to commit.

anton-vlasenko committed:
Add unit tests.

anton-vlasenko committed:
Short circuit the WP_REST_Search_Controller controller.

anton-vlasenko committed:
Fix the WP_Test_REST_Posts_Controller::test_get_items_pagination_headers() test.

anton-vlasenko committed:
Short circuit the WP_REST_Sidebars_Controller controller.

anton-vlasenko committed:
Remove non-existent test method parameter.

anton-vlasenko committed:
Add unit tests.

anton-vlasenko committed:
Fix backward compatibility issues in the WP_REST_Font_Collections_Controller for HEAD requests.

anton-vlasenko committed:
Short circuit the WP_REST_Templates_Controller controller.

anton-vlasenko committed:
Add unit tests.

anton-vlasenko committed:
Short circuit the WP_REST_Widget_Types_Controller controller.

anton-vlasenko committed:
Fix the unit test.

anton-vlasenko committed:
Add unit tests.

anton-vlasenko committed:
Short circuit the WP_REST_Widgets_Controller controller.

anton-vlasenko committed:
Add unit tests.

anton-vlasenko committed:
Unify inline comments.

janusqa committed:
Short circuit the WP_REST_Autosaves_Controller controller.

anton-vlasenko committed:
Add unit tests.

anton-vlasenko committed:
Short-circuit the WP_REST_Template_Autosaves_Controller and WP_REST_Template_Revisions_Controller controllers.

anton-vlasenko committed:
Rename test methods.

anton-vlasenko committed:
Add unit tests.

anton-vlasenko committed:
Add unit tests.

anton-vlasenko committed:
Short circuit the WP_REST_Pattern_Directory_Controller controller.

anton-vlasenko commented:
I believe this PR is ready for review.

Additionally, I've updated some REST controllers beyond the ones [listed](https://github.com/WordPress/wordpress-develop/pull/7970#issuecomment-2549749805), including:

- `WP_REST_Revisions_Controller`;
- `WP_REST_Template_Revisions_Controller`
- `WP_REST_Autosaves_Controller`;
- `WP_REST_Template_Autosaves_Controller`.

mukeshpanchal27 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-font-collections-controller.php:
```diff
@@ -97,11 +99,21 @@ public function get_items( $request ) {
 				continue;
 			}
+
+			/**
```
```suggestion
			/*
```

Nit-pick

anton-vlasenko replied:
I agree.
I’m accustomed to using DocBlock-style comments because they are correctly parsed by the PHP tokenizer. However, I acknowledge that this is not a true DocBlock comment since it doesn’t contain any PHPDoc tags. Therefore, using a regular multiline comment is more appropriate in this case.  

I’ve created an issue addressing a related problem (posting it here for context and reference): https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/780.

anton-vlasenko committed:
Update src/wp-includes/rest-api/endpoints/class-wp-rest-font-collections-controller.php

Don't use a PHPDoc block because it doesn't contain any PHPDoc tags.

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

anton-vlasenko commented:
~~I'm working on fixing the merge conflicts.~~

UPD: Done.

anton-vlasenko committed:
Fix CS.

anton-vlasenko committed:
Fix the failing PHPUnit tests.

spacedmonkey reviewed (approved):
Looks good to me
